### PR TITLE
Add share mail link feature

### PR DESCRIPTION
**Pull request by NicolasBourdin88**

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/data/models/ShareThread.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.mail.data.models
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+class ShareThread {
+    var url: String? = null
```
Comment (TommyDL-Infomaniak) [inline app/src/main/java/com/infomaniak/mail/data/models/ShareThread.kt:24]:
```suggestion
    var url: String
```

**Review by TommyDL-Infomaniak — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 2]
```diff
@@ -1061,6 +1063,19 @@ class MainViewModel @Inject constructor(
         return messageController.getMessageCountInThreadForFolder(threadUid, folderId, mailboxContentRealm()) == 1L
     }
 
+    fun shareThreadUrl(messageUid: String, activityContext: Context) {
+        val message = messageController.getMessage(messageUid) ?: return
+
+        viewModelScope.launch(ioCoroutineContext) {
+            val mailboxUuid = currentMailbox.value?.uuid!!
```
Comment (TommyDL-Infomaniak) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:1070]: Return if null

**Review by TommyDL-Infomaniak — CHANGES_REQUESTED** [review]

**Review by TommyDL-Infomaniak — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 3]
```diff
@@ -1061,6 +1063,19 @@ class MainViewModel @Inject constructor(
...
+
+        viewModelScope.launch(ioCoroutineContext) {
+            val response = ApiRepository.getShareLink(mailboxUuid, message.folderId, message.shortUid)
+
+            if (response.isSuccess() && response.data != null) {
+                response.data!!.url.let { activityContext.shareString(it) }
+            }
+        }
+    }
+
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:1078]: You have a code smell here. If message is `null`, you initialize a `val` for nothing.

```suggestion
fun shareThreadUrl(messageUid: String, activityContext: Context) {
        messageController.getMessage(messageUid)?.let { message ->
            val mailboxUuid = currentMailbox.value?.uuid ?: return

            viewModelScope.launch(ioCoroutineContext) {
                val response = ApiRepository.getShareLink(mailboxUuid, message.folderId, message.shortUid)

                if (response.isSuccess() && response.data != null) {
                    activityContext.shareString(response.data!!.url)
                }
            }
        }
    }
```

File: app/src/main/java/com/infomaniak/mail/data/models/ShareThread.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.mail.data.models
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ShareThread(
+    val url: String
+)
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/data/models/ShareThread.kt:25]:
```suggestion
data class ShareThread(val url: String)
```

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 5]
```diff
@@ -1061,6 +1063,19 @@ class MainViewModel @Inject constructor(
...
 
+    fun shareThreadUrl(messageUid: String, activityContext: Context) {
+        val message = messageController.getMessage(messageUid) ?: return
+        val mailboxUuid = currentMailbox.value?.uuid ?: return
+
+        viewModelScope.launch(ioCoroutineContext) {
+            val response = ApiRepository.getShareLink(mailboxUuid, message.folderId, message.shortUid)
+
+            if (response.isSuccess() && response.data != null) {
+                response.data!!.url.let { activityContext.shareString(it) }
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:1074]: What's the point of this `let` ?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt [hunk 6]
```diff
@@ -53,6 +53,7 @@ abstract class MailActionsBottomSheetDialog : ActionsBottomSheetDialog() {
         override fun onFavorite() = Unit
         override fun onReportJunk() = Unit
         override fun onPrint() = Unit
+        override fun onShare() = Unit
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MailActionsBottomSheetDialog.kt:56]: If we can keep the same order as in the interface:

```kotlin
    private var onClickListener: OnActionClick = object : OnActionClick {

        //region Actions
        override fun onArchive() = Unit
        override fun onReadUnread() = Unit
        override fun onMove() = Unit
        override fun onPostpone() = Unit
        override fun onFavorite() = Unit
        override fun onReportJunk() = Unit
        override fun onPrint() = Unit
        override fun onShare() = Unit
        override fun onReportDisplayProblem() = Unit
        //endregion

        //region Main actions
        override fun onReply() = Unit
        override fun onReplyAll() = Unit
        override fun onForward() = Unit
        override fun onDelete() = Unit
        //endregion
    }
```

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 7]
```diff
@@ -17,7 +17,9 @@
  */
 package com.infomaniak.mail.ui
 
+import android.app.Activity
 import android.app.Application
+import android.content.Context
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:22]: Also, you are in an `AndroidViewModel`. You already have the context inside the `ViewModel`. No need to import `Context` and also `Activity`

**Review by tevincent — CHANGES_REQUESTED** [review]

**Review by tevincent — APPROVED** [review]

**Pull request merged**